### PR TITLE
feature(typescript_indexer): index binding element parameters

### PR DESCRIPTION
**Pull request by ayazhafiz**
Currently we entirely skip over function parameters that are defined
with binding elements, like

```typescript
function foo({a, b: {c, d}}, [e, [f]]) { ... }
```

Implementing this is fairly straightforward. The existing parameter
visitor implementation is pulled out into its own method, and when we
encounter a set of parameter declarations they are recursively traversed
for their declarations in a binding expression.

Maybe ideally, these paramaters would be named `a: 0.0`, `c: 0.1.0`,
etc. to more accurately reflect how they are processed as arguments by
JavaScript/TypeScript. But Kythe doesn't support that and for the user
it makes more sense to think about them linearly.

### CODE REVIEW

**Comment by ayazhafiz** [comment]
cc @evmar

File: kythe/typescript/indexer.ts [hunk 1]
```diff
@@ -1006,7 +1006,7 @@ class Visitor {
 
   visitFunctionLikeDeclaration(decl: ts.FunctionLikeDeclaration) {
     this.visitDecorators(decl.decorators || []);
-    let kFunc: VName|undefined = undefined;
+    let kFunc: VName;
```
Comment (evmar) [inline kythe/typescript/indexer.ts]: Good change!

File: kythe/typescript/indexer.ts [hunk 2]
```diff
@@ -1117,7 +1086,76 @@ class Visitor {
...
+    }
+  }
+
+  /**
+   * Visits a function parameters, which can be recursive in the case of
+   * parameters created via bound elements:
+   *    function foo({a, b: {c, d}}, e, f) {}
+   * In this code, a, c, d, e, f are all parameters with increasing parameter
+   * numbers [0, 4]. Ideally the parameter number would be hierarchical, with
+   * `a` having a parameter number of 0.0, but Kythe doesn't support that.
```
Comment (evmar) [inline kythe/typescript/indexer.ts]: Why is it idea to have these specific names?  It only needs to be consistent within this module, right?

File: kythe/typescript/indexer.ts [hunk 3]
```diff
@@ -1117,7 +1086,76 @@ class Visitor {
...
+              }
+            default:
+              break;
+          }
+
+          if (ts.isParameter(param) && param.type) this.visitType(param.type);
+          if (param.initializer) this.visit(param.initializer);
+        }
+
+    for (const [, element] of toArray(parameters.entries())) {
```
Comment (evmar) [inline kythe/typescript/indexer.ts]: Isn't parameters already an array?  Why not
for (const element of parameters) {
?

File: kythe/typescript/indexer.ts (hunk 2 above)
Comment (ayazhafiz) [reply kythe/typescript/indexer.ts]: Oops! Removed.

File: kythe/typescript/indexer.ts (hunk 3 above)
Comment (ayazhafiz) [reply kythe/typescript/indexer.ts]: yeah, i dont know why i did this. thanks!

**Pull request merged**